components/kyma-adapter/internal/handler/handler.go

```diff
@@ -1,16 +1,300 @@
 package handler
 
 import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
 	"net/http"
+	"strings"
 
-	"github.com/machinebox/graphql"
+	"github.com/kyma-incubator/compass/components/kyma-adapter/internal/types"
+
+	"github.com/go-openapi/runtime/middleware/header"
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+	"github.com/kyma-incubator/compass/components/director/pkg/httputils"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/pkg/errors"
+)
+
+const (
+	assignOperation   = "assign"
+	unassignOperation = "unassign"
+
+	readyState         string = "READY"
+	configPendingState string = "CONFIG_PENDING"
+	createErrorState   string = "CREATE_ERROR"
+	deleteErrorState   string = "DELETE_ERROR"
+
+	successStatusCode int = http.StatusOK
 )
 
-// AdapterHandler processes received requests
+// Client is responsible for making internal graphql calls to the director
+//go:generate mockery --name=Client --output=automock --outpkg=automock --case=underscore --disable-version-string
+type Client interface {
+	GetApplicationBundles(ctx context.Context, appID, tenant string) ([]*graphql.BundleExt, error)
+	CreateBasicBundleInstanceAuth(ctx context.Context, tenant, bndlID, rtmID, username, password string) error
+	CreateOauthBundleInstanceAuth(ctx context.Context, tenant, bndlID, rtmID, tokenServiceURL, clientID, clientSecret string) error
+	UpdateBasicBundleInstanceAuth(ctx context.Context, tenant, authID, bndlID, username, password string) error
+	UpdateOauthBundleInstanceAuth(ctx context.Context, tenant, authID, bndlID, tokenServiceURL, clientID, clientSecret string) error
+	DeleteBundleInstanceAuth(ctx context.Context, tenant, authID string) error
+}
+
+// AdapterHandler is the Kyma Tenant Mapping Adapter handler which processes the received requests
 type AdapterHandler struct {
-	DirectorGqlClient *graphql.Client
+	DirectorGqlClient Client
+}
+
+// NewHandler creates an AdapterHandler
+func NewHandler(directorGqlClient Client) *AdapterHandler {
+	return &AdapterHandler{DirectorGqlClient: directorGqlClient}
 }
 
 // HandlerFunc is the implementation of AdapterHandler
 func (a AdapterHandler) HandlerFunc(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
```

Comment:
The handlerFunc can be simplified. The assign and unassign flows have a lot in common and can be merged. The HandlerFunc can look like:

- decode and validate request
- get bundles for application
- if no bundles - respond success
- determine the function that should be called and build input for it based on the operation and the presence of Instance auths.
  - check if there is Instance auth for any of the bundles - IAExists
  - determine the operation - Operation
  - IAExists == false  and operation == unassign - then nothing to do, IAExists==true and operation = assign - then update Instance auth
  - Prepare struct with input based on the flow
 - for each bundle execute the function with the input

That way the handler will be more readable also there will be no code duplication and no need for duplicating fixes etc